File: mergeSort.js
//Merge sort algorithm - From: https://medium.com/javascript-in-plain-english/javascript-merge-sort-3205891ac060

function mergeSort(arr) {
    if (arr.length <= 1) {
        return arr;
    }

    const mid = Math.floor(arr.length / 2);
    const left = arr.slice(0, mid);
    const right = arr.slice(mid);

    return merge(
        mergeSort(left), mergeSort(right)
    );
}

function merge(left, right) {
    let res = [];
    let leftIdx = 0;
    let rightIdx = 0;

    while (leftIdx < left.length && rightIdx < right.length) {
        const firstDistance = getDistanceFromLatLonInKm(left[leftIdx].lat, left[leftIdx].lon);
        const secondDistance = getDistanceFromLatLonInKm(right[rightIdx].lat, right[rightIdx].lon);
        if (firstDistance < secondDistance) {
            res.push(left[leftIdx]);
            leftIdx++;
        } else {
            res.push(right[rightIdx]);
            rightIdx++;
        }
    }

    return res.concat(left.slice(leftIdx)).concat(right.slice(rightIdx));
}

File: aspnet/Examen/MergeSort.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

//Merge sort algorithm - From: https://medium.com/javascript-in-plain-english/javascript-merge-sort-3205891ac060 - Ported to C# by me

namespace Examen
{
    public class MergeSort
    {
        public List<GPSData> result = new List<GPSData>();
        public List<List<GPSData>> queue = new List<List<GPSData>>();
        Haversine haversine = new Haversine();

        public List<GPSData> Sort(List<GPSData> unsorted)
        {
            if (unsorted.Count <= 1)
            {
                return unsorted;
            }

            int mid = unsorted.Count / 2;
            List<GPSData> left = unsorted.Take(mid).ToList();
            List<GPSData> right = unsorted.Skip(mid).ToList();

            return Merge(Sort(left), Sort(right));
        }

        public List<GPSData> Merge(List<GPSData> left, List<GPSData> right)
        {
            List<GPSData> res = new List<GPSData>();
            int leftIdx = 0;
            int rightIdx = 0;

            while (leftIdx < left.Count && rightIdx < right.Count)
            {
                double firstDistance = haversine.getDistance(left.First().lat, left.First().lon);
                double secondDistance = haversine.getDistance(right.First().lat, right.First().lon);
                if (firstDistance < secondDistance)
                {
                    res.Add(left[leftIdx]);
                    leftIdx++;
                } else
                {
                    res.Add(right[rightIdx]);
                    rightIdx++;
                }
            }

            res.AddRange(left.Skip(leftIdx).ToList());
            res.AddRange(right.Skip(rightIdx).ToList());
            return res;

        }
    }
}
